vignettes/parallel.Rmd

```diff
@@ -58,3 +58,96 @@ There is some overhead associated with running tests in parallel:
+
+-   Reporters that stop after a certain number of failures can only stop at the end of a test file.
+
+-   Reporters report all information about a file at once, unless they support *parallel updates*.
+    E.g.
+    `ProgressReporter` does not update its display until a test file is complete.
+
+-   The standard output and standard error, i.e. `print()`, `cat()`, `message()`, etc. output from the test files are lost currently.
+    If you want to use `cat()` or `message()` for print-debugging test cases, then the best is to temporarily run tests sequentially, by changing the `Config` entry in `DESCRIPTION` or selecting a non-parallel reporter, e.g. the `CheckReporter`:
+
+    ``` {.r}
+    devtools::test(filter = "badtest", repoter = "check")
+    ```
+
+### Writing parallel reporters
+
+To support parallel tests, a reporter must be able to function when the test files run in a subprocess.
+For example `DebugReporter` does not support parallel tests, because it requires direct interaction with the frames in the subprocess.
+When running in parallel, testthat does not provide location information (source references) for test successes.
+
+To support parallel tests, a reporter must set `self$capabilities$parallel_support` to `TRUE` in its `initialize()` method:
+
+``` {.r}
+...
+initialize = function(...) {
+  super$initialize(...)
+  self$capabilities$parallel_support <- TRUE
+  ...
+}
+...
+```
+
+When running in parallel, testthat runs the reporter in the main process, and relays information between the reporter and the test code transparently.
+(Currently the reporter does not even know that the tests are running in parallel.)
+
+If a reporter does not support parallel updates (see below), then testthat internally caches all calls to the reporter methods from subprocesses, until a test file is complete.
+This is because these reporters are not prepared for running multiple test files concurrently.
+Once a test file is complete, testthat calls the reporter's `$start_file()` method, relays all `$start_test()` , `$end_test()`, `$add_result()`, etc. calls in the order they came in from the subprocess, and calls `$end_file()` .
+
+### Parallel updates
+
+The `ParallelProgressReporter` supports parallel updates.
+This means that once a message from a subprocess comes in, the reporter is updated immediately.
+For this to work, a reporter must be able to handle multiple test files concurrently.
+
+A reporter declares parallel update support by setting `self$capabilities$parallel_updates` to `TRUE`:
+
+``` {.r}
+...
+initialize = function(...) {
+  super$initialize(...)
+  self$capabilities$parallel_support <- TRUE
+  self$capabilities$parallel_updates <- TRUE
+  ...
+}
+...
+```
+
+For these reporters, testthat does not cache the messages from the subprocesses.
+Instead, when a message comes in:
+
+-   It calls the `$start_file()` method, letting the reporter know which file the following calls apply to.
```

Comment:
Is there any worry about interleaving of `$start_file()` or the following `$start_test()` / `$add_result()` from multiple sub-processes? How is synchronization handled there? (seems like in theory the parent process could receive multiple `$start_file()` messages at once?)

Reply:
No, there is no such worry, because the subsequent `start_file()` messages are not real, they are not coming from the subprocess. The subprocess only sends one `start_file()` at the beginning. 

The logic in the main process prepends a `start_file()` before each subsequent message from the subprocess, to let the reporter know that the next message(s) are for a certain file.